File: service/internal/infrastructure/health/endpoints.go
package health

import (
	"net/http"

	"github.com/go-chi/chi"
	"github.com/sazzer/mire/service/internal/utils/httpresponse"
)

// componentHealthModel represents the HTTP Model for the health of a single component.
type componentHealthModel struct {
	Status  string `json:"status"`
	Message string `json:"message,omitempty"`
}

// systemHealthModel represents the HTTP Model for the health of the entire system.
type systemHealthModel struct {
	Status     string                          `json:"status"`
	Components map[string]componentHealthModel `json:"components"`
}

// endpoints is a struct that contains all of the HTTP Endpoints for checking the health of the system.
type endpoints struct {
	service Service
}

// getHealth is the endpoint for getting the health of the system.
func (e *endpoints) getHealth(w http.ResponseWriter, r *http.Request) {
	systemHealth := e.service.CheckHealth(r.Context())

	var response systemHealthModel
	response.Components = map[string]componentHealthModel{}

	for componentName, componentHealth := range systemHealth.Components {
		if componentHealth.Error == nil {
			response.Components[componentName] = componentHealthModel{
				Status: "HEALTHY",
			}
		} else {
			response.Components[componentName] = componentHealthModel{
				Status:  "UNHEALTHY",
				Message: componentHealth.Error.Error(),
			}
		}
	}

	response.Status = "HEALTHY"
	statusCode := http.StatusOK
	if systemHealth.Status() == StatusUnhealthy {
		response.Status = "UNHEALTHY"
		statusCode = http.StatusServiceUnavailable
	}

	httpresponse.RenderJSON(w, statusCode, response)
}

// RegisterRoutes is the implementation of the server.Configurer interface to register these
// endpoints with the HTTP Server.
func (e endpoints) RegisterRoutes(r chi.Router) {
	r.Get("/health", e.getHealth)
}

File: service/crates/health/src/endpoints.rs
use super::{ComponentHealth, HealthService, SystemHealth};
use actix_http::{http::StatusCode, Error, Response};
use actix_web::{get, web::Data, HttpRequest, HttpResponse, Responder};
use futures::future::{ok, Ready};
use serde::Serialize;
use std::collections::HashMap;

/// HTTP Handler for checking the health of the system.
///
/// # Parameters
/// - `health_service` - The Health Service to get the system health with
///
/// # Returns
/// The HTTP Model for the response
#[get("/health")]
#[tracing::instrument(
    fields(http_method = "GET", http_path = "/health"),
    skip(health_service)
)]
pub async fn get_health(health_service: Data<HealthService>) -> impl Responder {
    let system_health = health_service.check_health().await;

    SystemHealthModel::from(system_health)
}

/// HTTP Representation of the health of a single component
#[derive(Serialize)]
struct ComponentHealthModel {
    /// Whether the component is healthy or not
    pub healthy: bool,
    /// A message about the component if it was unhealthy
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
}

impl From<ComponentHealth> for ComponentHealthModel {
    /// Convert the internal `ComponentHealth` instance into the HTTP Response.
    fn from(health: ComponentHealth) -> Self {
        ComponentHealthModel {
            healthy: health.is_healthy(),
            message: match health {
                ComponentHealth::Healthy => None,
                ComponentHealth::Unhealthy(e) => Some(e),
            },
        }
    }
}

/// HTTP Representation of the health of the entire system
#[derive(Serialize)]
struct SystemHealthModel {
    /// Whether the system is healthy or not
    pub healthy: bool,
    /// The individual components that make up the system.
    pub components: HashMap<String, ComponentHealthModel>,
}

impl From<SystemHealth> for SystemHealthModel {
    /// Convert the internal `SystemHealth` instance into the HTTP Response.
    fn from(health: SystemHealth) -> Self {
        SystemHealthModel {
            healthy: health.is_healthy(),
            components: health
                .components
                .into_iter()
                .map(|(name, component)| (name, component.into()))
                .collect(),
        }
    }
}

impl Responder for SystemHealthModel {
    type Error = Error;
    type Future = Ready<Result<Response, Error>>;

    /// Generate an HTTP Response for the System Health.
    ///
    /// The status code is either `StatusCode::OK` for a healthy system or `StatusCode::SERVICE_UNAVAILABLE`
    /// for an unhealthy system, and the body is the JSON version of this type.
    fn respond_to(self, _: &HttpRequest) -> Self::Future {
        let status_code = if self.healthy {
            StatusCode::OK
        } else {
            StatusCode::SERVICE_UNAVAILABLE
        };

        let response = HttpResponse::build(status_code).json(self);

        ok(response)
    }
}
